Performance Timeline API

ferjm opened the pull request:
[Performance Timeline API](https://www.w3.org/TR/performance-timeline-2/) implementation.

This API is required to allow DOM access to the [Paint Timing API](https://wicg.github.io/paint-timing/#example) metrics implemented in #17256. Unfortunately, I couldn't test it properly, as its usage depends on other APIs like the Paint Timing, User Timing, Resource Timing or Server Timing APIs. I'll work in the integration with the Paint Timing API next.

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] There are [WPTs](https://github.com/servo/servo/tree/master/tests/wpt/web-platform-tests/performance-timeline) for this API, however they depend on the implementation of the User Timing and the Resource Timing APIs, which I'll hopefully be implementing soon.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/18028)
<!-- Reviewable:end -->

ferjm committed:
Performance Timeline API

highfive commented:
Heads up! This PR modifies the following files:
 * @fitzgen: components/script/dom/webidls/PerformanceObserverEntryList.webidl, components/script/dom/performance.rs, components/script/dom/webidls/PerformanceObserver.webidl, components/script/script_thread.rs, components/script/dom/performanceobserver.rs and 9 more
 * @kichjang: components/script/dom/webidls/PerformanceObserverEntryList.webidl, components/script/dom/performance.rs, components/script/dom/webidls/PerformanceObserver.webidl, components/script/script_thread.rs, components/script/dom/performanceobserver.rs and 9 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

ferjm commented:
r? @jdm

jdm commented on components/script/script_thread.rs:
```diff
@@ -269,6 +270,8 @@ pub enum MainThreadScriptMsg {
     /// Notifies the script thread that a new worklet has been loaded, and thus the page should be
     /// reflowed.
     WorkletLoaded(PipelineId),
+    /// XXX
```
Tasks that originate from the performance timeline task source.

jdm commented on components/script/dom/webidls/PerformanceObserverEntryList.webidl:
```diff
@@ -0,0 +1,15 @@
+ * You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ * The origin of this IDL file is
+ * https://w3c.github.io/performance-timeline/#the-performanceobserverentrylist-interface
```
https://w3c.github.io/performance-timeline/#performanceobserverentrylist-interface

jdm commented on components/script/dom/performanceobserverentrylist.rs:
```diff
@@ -0,0 +1,55 @@
+impl PerformanceObserverEntryListMethods for PerformanceObserverEntryList {
+    // https://www.w3.org/TR/performance-timeline-2/#dom-performanceobserver
+    fn GetEntries(&self)
+        -> Vec<Root<PerformanceEntry>> {
```
nit: move this to the previous line.

jdm commented on components/script/dom/performanceobserverentrylist.rs:
```diff
@@ -0,0 +1,55 @@
+}
+
+impl PerformanceObserverEntryListMethods for PerformanceObserverEntryList {
+    // https://www.w3.org/TR/performance-timeline-2/#dom-performanceobserver
```
Use https://w3c.github.io/performance-timeline/ for these links instead of the TR.

jdm commented on components/script/dom/performanceobserver.rs:
```diff
@@ -0,0 +1,107 @@
+    reflector_: Reflector,
+    #[ignore_heap_size_of = "can't measure Rc values"]
+    callback: Rc<PerformanceObserverCallback>,
+    global: JS<GlobalScope>,
```
We can always use `self.global()` instead of storing this.

jdm commented on components/script/dom/performanceobserver.rs:
```diff
@@ -0,0 +1,107 @@
+}
+
+impl PerformanceObserverMethods for PerformanceObserver {
+    // https://www.w3.org/TR/performance-timeline-2/#dom-performanceobserver-observe
```
https://w3c.github.io/performance-timeline/#dom-performanceobserver-observe()

jdm commented on components/script/dom/performanceobserver.rs:
```diff
@@ -0,0 +1,107 @@
+
+    // https://www.w3.org/TR/performance-timeline-2/#dom-performanceobserver-disconnect
+    fn Disconnect(&self) {
+        self.global.as_window().Performance().remove_observer(&self);
```
nit: no need for &.

jdm commented on components/script/dom/performanceobserver.rs:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        let performance = self.global.as_window().Performance();
+        performance.add_observer(&self, entry_types);
```
nit: no need for &.

jdm commented on components/script/dom/performanceentry.rs:
```diff
@@ -0,0 +1,76 @@
+}
+
+impl PerformanceEntryMethods for PerformanceEntry {
+    // https://www.w3.org/TR/performance-timeline-2/#dom-performanceentry-name
```
https://w3c.github.io/performance-timeline/#dom-performanceentry-name

jdm commented on components/script/dom/performance.rs:
```diff
@@ -41,6 +103,79 @@ impl Performance {
+
+    fn notify_observers(&self) {
+        self.pending_notification_observers_task.set(false);
+        let _ = self.observers.borrow().iter().map(|o| o.observer.notify());
```
Is there a risk of an observer's callback modifying self.observers?

ferjm replied:
Yes, the observer can call `observer.disconnect()` for example. Which would end up modifying `self.observers`. Should I wrap `self.observers` with a `RwLock` or such?

jdm replied:
According to https://w3c.github.io/performance-timeline/#queue-a-performanceentry we should be operating on a copy of the list of observers in order to avoid this problem.

jdm commented on components/script/dom/performance.rs:
```diff
@@ -41,6 +103,79 @@ impl Performance {
+    pub fn add_observer(&self,
+                        observer: &DOMPerformanceObserver,
+                        entry_types: Vec<DOMString>) {
+        self.observers.borrow_mut().push(PerformanceObserver {
```
According to https://w3c.github.io/performance-timeline/#dom-performanceobserver-observe() we should be replacing any existing matching one.

jdm commented on components/script/dom/performance.rs:
```diff
@@ -2,22 +2,81 @@
+            match entry_type {
+                Some(ref entry_type) => *e.name() == name && *e.entry_type() == *entry_type,
+                None => *e.name() == name
+            }
```
```rust
filter(|e| {
    *e.name() == name &&
    entry_type.as_ref().map_or(true, |type_| *e.entry_type() == type_)
})
```

jdm reviewed (changes requested):
This was a straightforward review; thanks!

jdm commented on components/script/script_thread.rs:
```diff
@@ -269,6 +270,8 @@ pub enum MainThreadScriptMsg {
     /// Notifies the script thread that a new worklet has been loaded, and thus the page should be
     /// reflowed.
     WorkletLoaded(PipelineId),
+    /// Task that originate from the performnace timeline task source.
```
Tasks

jdm commented on components/script/dom/performance.rs:
```diff
@@ -41,6 +101,86 @@ impl Performance {
+        let _ = self.observers.borrow()
+                              .iter()
+                              .filter(|o| o.entry_types.contains(entry.entry_type()))
+                              .map(|o| o.observer.queue_entry(entry));
```
Iterators are lazy, so the entries will never be queued here. It would be better to write this as:
```rust
for o in self.observers.borrow().iter().filter(|o| o.entry_types.contains(entry.entry_type()))) {
    o.observer.queue_entry(entry);
}
```

jdm commented on components/script/dom/performance.rs:
```diff
@@ -41,6 +101,86 @@ impl Performance {
+
+    fn main_thread_handler(self: Box<NotifyPerformanceObserverRunnable>,
+                           _: &ScriptThread) {
+        (*self).owner.root().notify_observers();
```
Does `self` work instead of `(*self)`?

jdm reviewed (changes requested):
Sorry, I noticed a couple things that looked suspicious. It might be valuable to create a test API that verifies that this actually works as expected.

ferjm commented:
Thank you @jdm! I wrote the integration with the Paint Timing API to test this patch and it's working as expected after addressing your feedback. https://github.com/ferjm/servo/commit/2042941a3edf0cabc22aee4add1049e61e892563

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 79198f5 has been approved by `jdm`

<!-- @bors-servo r=jdm 79198f572e8c1e4853b23edd8c1444530081b90b -->

bors-servo commented:
:hourglass: Testing commit 79198f572e8c1e4853b23edd8c1444530081b90b with merge 2000ef818a96b645af8439b0eac0c60122378be1...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/8342)

ferjm commented:
@bors-servo retry

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5491)

jdm commented:
```

  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.worker.html:
  │ FAIL [expected PASS] Untitled
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface PerformanceEntry to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
  │ 
  │ test_interfaces/<@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:73:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ test_interfaces@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:2:3
  └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.worker.js:9:1

  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.html:
  │ FAIL [expected PASS] Interfaces exposed on the window
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface PerformanceEntry to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
  │ 
  │ test_interfaces/<@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:73:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ test_interfaces@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:2:3
  └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:13:1

  ▶ TIMEOUT [expected OK] /navigation-timing/nav2_test_document_replaced.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1
```

ferjm commented:
I have updated the `interfaces.html` test and the test expectation for `navigation-timing/nav2_test_document_replaced.html.ini`. This test used to fail because `window.performance.getEntriesByType()` wasn't defined. With the current patch, this function is now defined (although "navigation" is not a valid entry type). However the `onload` event for the [3rd test step](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/navigation-timing/nav2_test_document_replaced.html#L37) is not being fired as expected, which causes the test to timeout. There seems to be something wrong with history, bfcache and onload events. @jdm are these known issues? Otherwise, I'll try to get a minimal test case for this and file a new issue.

ferjm commented:
@bors-servo retry

jdm commented:
I would not be surprised if we don't get load events right for traversing history, but I don't know of an existing issue for it. Please file one!

@bors-servo: r+

bors-servo commented:
:pushpin: Commit 233d293 has been approved by `jdm`

<!-- @bors-servo r=jdm 233d29330ce7f1675d0bbf91a173acd4ee8846f6 -->

bors-servo commented:
:hourglass: Testing commit 233d29330ce7f1675d0bbf91a173acd4ee8846f6 with merge c15ed90390271d390418c3753bd3ddbfcb4c9076...

jdm commented:
@bors-servo: r-
Need to run `./mach update-manifest` and commit the changes.

jdm commented:
You will need to update interfaces.worker.js as well.

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/8693)

ferjm commented:
@bors-servo retry

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 52348f1 has been approved by `jdm`

<!-- @bors-servo r=jdm 52348f1fccea3e4abc2ad80d001a43f5f51cbdc8 -->

bors-servo commented:
:hourglass: Testing commit 52348f1fccea3e4abc2ad80d001a43f5f51cbdc8 with merge 6988c7424ddfdd9a98e6a458ff2ad307a74237aa...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8374), [arm32](http://build.servo.org/builders/arm32/builds/8598), [arm64](http://build.servo.org/builders/arm64/builds/8630), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8714), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5481), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5518), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8714), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1190), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1193), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5410), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5470), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1192), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1197), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3961)
Approved by: jdm
Pushing 6988c7424ddfdd9a98e6a458ff2ad307a74237aa to master...
